src/linker/Linker/LinkContext.cs

```diff
@@ -663,6 +663,169 @@ public TargetRuntimeVersion GetTargetRuntimeVersion ()
+			return md;
+		}
+
+		public MethodDefinition TryResolveMethodDefinition (MethodReference methodReference)
+		{
+			if (methodReference is MethodDefinition md)
+				return md;
+
+			if (methodReference is null)
+				return null;
+
+			if (methodresolveCache.TryGetValue (methodReference, out md))
+				return md;
+
+			md = methodReference.Resolve ();
+			methodresolveCache.Add (methodReference, md);
+			return md;
+		}
+
+		public FieldDefinition ResolveFieldDefinition (FieldReference fieldReference)
+		{
+			if (fieldReference is FieldDefinition fd)
+				return fd;
+
+			if (fieldReference is null)
+				return null;
+
+			if (fieldresolveCache.TryGetValue (fieldReference, out fd)) {
+				if (fd == null && !IgnoreUnresolved)
+					ReportUnresolved (fieldReference);
+
+				return fd;
+			}
+
+			fd = fieldReference.Resolve ();
+			if (fd == null && !IgnoreUnresolved) {
+				ReportUnresolved (fieldReference);
+			}
+
+			fieldresolveCache.Add (fieldReference, fd);
+			return fd;
+		}
+
+		public FieldDefinition TryResolveFieldDefinition (FieldReference fieldReference)
+		{
+			if (fieldReference is FieldDefinition fd)
+				return fd;
+
+			if (fieldReference is null)
+				return null;
+
+			if (fieldresolveCache.TryGetValue (fieldReference, out fd))
+				return fd;
+
+			fd = fieldReference.Resolve ();
+			fieldresolveCache.Add (fieldReference, fd);
+			return fd;
+		}
+
+		public TypeDefinition ResolveTypeDefinition (TypeReference typeReference)
+		{
+			if (typeReference is TypeDefinition td)
+				return td;
+
+			if (typeReference is null)
+				return null;
+
+			if (typeresolveCache.TryGetValue (typeReference, out td)) {
+				if (td == null && !IgnoreUnresolved)
+					ReportUnresolved (typeReference);
+
+				return td;
+			}
+
+			//
+			// Types which never have TypeDefinition or can have ambiguous definition should not be passed in
+			//
+			if (typeReference is GenericParameter || (typeReference is TypeSpecification && typeReference is not GenericInstanceType))
+				throw new NotSupportedException ($"TypeDefinition cannot be resolved from '{typeReference.GetType ()}' type");
+
+			td = typeReference.Resolve ();
+			if (td == null && !IgnoreUnresolved) {
+				ReportUnresolved (typeReference);
+			}
+
+			typeresolveCache.Add (typeReference, td);
+			return td;
+		}
+
+		public TypeDefinition TryResolveTypeDefinition (TypeReference typeReference)
+		{
+			if (typeReference is TypeDefinition td)
+				return td;
+
+			if (typeReference is null || typeReference is GenericParameter)
+				return null;
+
+			if (typeresolveCache.TryGetValue (typeReference, out td))
+				return td;
+
+			if (typeReference is TypeSpecification ts) {
+				if (typeReference is FunctionPointerType) {
+					td = null;
+				} else {
+					//
+					// It returns element-type for arrays and also element type for wrapping types like ByReference, PinnedType, etc
+					//
+					td = TryResolveTypeDefinition (ts.GetElementType ());
+				}
```

Comment:
Should we create an issue to handle at least some of these at the callsite - resolving to element-type is dangerous. Not counting that it also hides generic instantiations.

Reply:
I actually experimented with it a bit and I think this logic is fine. Most code still needs to resolve to the element so the separation would not have benefits beyond perhaps type system purity.